### ISSUE TITLE
Clarifications about metrics calculations

### ISSUE THREAD

**Issue by snexus** [edtechre/pybroker#7]
Hi,

Trying to reconcile how final metrics are calculated - how `end_market_value` is calculated and what is its relationship to PnL and unrealized PnL? 

For example in the screenshot below, the final market value of the long-only portfolio is much smaller than realized + unrealized PnL:

![image](https://user-images.githubusercontent.com/1496235/224484020-eef01cd0-9f56-4451-892a-aeef4019c506.png)

**Comment by edtechre** [comment]
Thank you for bringing this to my attention, snexus.

Currently, PnL is calculated as the aggregate PnL of all trades executed so far. It does not take into account the current market value of the portfolio, but only the trades that have already been executed.

After reviewing this, I agree that this can be confusing. It would be clearer to calculate PnL as the difference between the market value and the initial value of the portfolio. Would this be helpful for you?

**Comment by snexus** [comment]
Thanks for the clarifications about trades, @edtechre 

I suspect there is a potential problem in the **market_value** and **equity** calculation, not the PnL. 

From my understanding, if all the money is reinvested, the total value of the portfolio should reconcile to **cash + realised PnL + initial value**.  If the initial value is 100,000, the total value of the portfolio from the above screenshot should evaluate to 100,000 + 550,065.66 + 13,534.28?

As a piece of additional information, the portfolio code looks like this: 
I am using signals `is_inside` from external algo to decide when to buy.

```python
N = 20
def exec_fn(ctx):
    if ctx.is_inside[-1] == 1:
        pos = ctx.long_pos()
        if not pos:
            ctx.buy_shares = ctx.calc_target_shares(1.0/N)
            
    elif ctx.is_inside[-1] == 0:
        pos = ctx.long_pos()
        if pos:
            ctx.sell_shares = pos.shares

df = (pd
          .read_csv("../data/predictions/predictions_2005_2020_3_20.csv.gz")
          .assign(date = lambda df: pd.to_datetime(df['date']))
          .loc[:,['date','symbol','open','high','low','close','is_inside']]
    )
pybroker.register_columns('is_inside')
tickers = df['symbol'].unique()

strateg_config = StrategyConfig(initial_cash=100000)
strategy = Strategy(df, start_date=df['date'].min(), end_date=df['date'].max(), config=strateg_config)
strategy.add_execution(exec_fn, tickers)
result = strategy.backtest(calc_bootstrap=False)

```

**Comment by edtechre** [comment]
Hi snexus,

I believe this is consistent with what I said earlier. The market value will be recomputed every bar using the close price of each open position, along with the current cash holding of the portfolio. The PnL you see is not recalculated every bar based on close price or equity of the portfolio, but is a running realized PnL based on all of the trades that were made. If the market value of the portfolio drops, it will not change the PnL until a trade is made that changes the position holdings. And the unrealized PnL is the current unrealized PnL of all of the open positions (not a running total).

I agree that this is confusing. I have a change ready to make this a single PnL value for realized + unrealized, where PnL would be the difference between the current market value and the initial value of the portfolio. I can commit and push this today. Would that work for you?

**Comment by snexus** [comment]
Hi @edtechre 

I understand that PnL is not recalculated every bar, and fine with that (though your proposed change will make it cleaner, thanks). 

My main confusion is the **final line** which should contain the total aggregated metrics of the portfolio. The point I am trying to make is that all realized + unrealized PnL from all previous trades + initial value + cash of the portfolio should be equal to the total equity, right? But I can't see it mathematically from the last line.

**Comment by edtechre** [comment]
Hi @snexus,

No, the last line will not be what you expect if there was no final trade made by the last bar of data, where a trade is defined as an entry and then an exit (i.e. buy then sell shares). PyBroker will not automatically liquidate the position by the last bar, but I will add an option for that (see https://github.com/edtechre/pybroker/issues/6).

For the unrealized + realized PnL, the difference between [initial_market_value](https://www.pybroker.com/en/latest/reference/pybroker.eval.html#pybroker.eval.EvalMetrics.initial_market_value) and [end_market_value](https://www.pybroker.com/en/latest/reference/pybroker.eval.html#pybroker.eval.EvalMetrics.end_market_value) will be equal to that in [TestResult#metrics](https://www.pybroker.com/en/latest/reference/pybroker.strategy.html#pybroker.strategy.TestResult.metrics). Note that total_pnl shown there will also be realized PnL from trades that were made (an entry + exit).

**Comment by edtechre** [comment]
Commit https://github.com/edtechre/pybroker/commit/b33fc3b39574585df83acfc2648ad8e5db3ba5a1 changes updates the portfolio to be a running value of current market value - initial portfolio value.

To summarize:

- The PnL now recorded in [TestResult#portfolio](https://www.pybroker.com/en/latest/reference/pybroker.strategy.html#pybroker.strategy.TestResult.portfolio) will reflect unrealized + realized PnL as being current market value - initial market value.
- [EvalMetrics#total_pnl](https://www.pybroker.com/en/latest/reference/pybroker.eval.html#pybroker.eval.EvalMetrics.total_pnl) will contain the realized PnL. That is, the PnL only for trades that were completed (an entry and exit)
- To also make things clearer, I will add a total_unrealized_pnl metric to [EvalMetrics](https://www.pybroker.com/en/latest/reference/pybroker.eval.html#pybroker.eval.EvalMetrics) to hold any difference between (end_market_value - initial_market_value) and the total_pnl.
- As mentioned, I will add a configuration option to automatically liquidate any open positions on the last bar of data, which will force a final trade to be realized. https://github.com/edtechre/pybroker/issues/6

Sorry for the headache!

**Comment by snexus** [comment]
Hi @edtechre 

Thanks for the clarifications. I understand now about the final trade, but it's value is captured by unrealized PnL, right?

We both agree on this:
> For the unrealized + realized PnL, the difference between [initial_market_value](https://www.pybroker.com/en/latest/reference/pybroker.eval.html#pybroker.eval.EvalMetrics.initial_market_value) and [end_market_value](https://www.pybroker.com/en/latest/reference/pybroker.eval.html#pybroker.eval.EvalMetrics.end_market_value) will be equal to that in [TestResult#metrics](https://www.pybroker.com/en/latest/reference/pybroker.strategy.html#pybroker.strategy.TestResult.metrics).

However, this equality doesn't correspond to what we are seeing in the last line of the result

Attaching a screenshot of the calculation using TestResult.metrics, it reconciles to the last line in the dataframe.

What we expect to see is:
end_market_value - initial_market_value = PnL + Unrealized PnL

But,

end_market_value - initial_market_value = 231411.46
total_pnl + unrealized_pnl = 550065.66 + 11759.03

![image](https://user-images.githubusercontent.com/1496235/224518661-fa5b3165-e363-4d7c-a6ae-323a8dcb128e.png)

**Comment by edtechre** [comment]
I have removed unrealized_pnl from the portfolio DataFrame. But can you attach your trades and orders captured from [TestResult](https://www.pybroker.com/en/latest/reference/pybroker.strategy.html#pybroker.strategy.TestResult)?

Also, feel free to email them to me if you don't want to attach them here.

**Comment by edtechre** [comment]
Thank you for your patience @snexus 

I can help verify that the PnL is correct for your trades. But what could be going on is:

- The market value is updated for each unique date in the backtest data.
- But the market value is only calculated using close prices recorded on that date. if a close price is missing, then the position is omitted from the calculation.
- This is a problem if your universe of symbols doesn't share the same timestamps.

This is a bug and the market value should instead be calculated from the last close price seen for each symbol.

**Comment by snexus** [comment]
Thanks for looking into this, @edtechre. I've sent you the trade and order results. Happy to supply the raw data + portfolio notebook as well, if needed.

**Comment by snexus** [comment]
> The market value is updated for each unique date in the backtest data.
> But the market value is only calculated using close prices recorded on that date. If a close price is missing, then the position is omitted from the calculation.
> This is a problem if your universe of symbols don't share the same timestamps.

Good points just checked my universe, and except for the few delisted companies the dates are equally spaced (excluding weekends) and contain all trading dates for all symbols. Will send you the universe by email as well. Thank you!

**Comment by edtechre** [comment]
Hi @snexus,

I verified your trades and the PnL amounts look good to me. After making a fix for the issue above, I see a different ending market value being reported, which is greater than the initial balance + total PnL. The unrealized PnL that was calculated would have been affected by the same issue as well.

I will commit the fix soon and push a new version. Thank you for your help looking into this!

**Comment by snexus** [comment]
Thanks @edtechre

**Comment by edtechre** [comment]
Committed fix in https://github.com/edtechre/pybroker/commit/153340c72f7ab8bfa94ffe3523ee828400c9ffd9

**Comment by edtechre** [comment]
Fixed in [v1.0.19](https://github.com/edtechre/pybroker/releases/tag/v1.0.19). Let me know if you need anything else.

**Comment by snexus** [comment]
Hi @edtechre - the total equity and PnL reconcile now, thanks for fixing that! Cross checked it with my own backtest outside of PyBroker - it looks good!
